src/graph_visualization.cpp

```diff
@@ -0,0 +1,550 @@
+                    nodes[i].pred.erase(nodes[i].pred.begin() + j);
+                    break;
+                }
+            }
+            break;
+        }
+    }
+
+    nodes.push_back(dummy);
+}
+
+void add_dummy_nodes(std::vector<tree_node> &nodes, std::vector<std::pair<tree_node::node_id, tree_node::node_id>> &deleted_edges) {
+    int flag = 1;
+    while (flag) {
+        flag = 0;
+        for (int i = 0; i < nodes.size(); i++) {
+            for (int j = 0; j < nodes[i].succ.size(); j++) {
+                tree_node succ_node = get_tree_node_by_id(nodes, nodes[i].succ[j]);
+                if (nodes[i].layer + 1 < succ_node.layer) {
+                    add_dummy_node(nodes, nodes[i], succ_node, 1);
+                    flag = 1;
+                    break;
+                }
+            }
+            if (flag == 1) {
+                break;
+            }
+        }
+    }
+
+    for (int i = 0; i < deleted_edges.size(); i++) {
+        tree_node first_node = get_tree_node_by_id(nodes, deleted_edges[i].first);
+        tree_node second_node = get_tree_node_by_id(nodes, deleted_edges[i].second);
+        
+        if (first_node.layer == second_node.layer + 1) {
+            for (int j = 0; j < nodes.size(); j++) {
+                if (nodes[j].id == first_node.id) {
+                    nodes[j].succ.push_back(second_node.id);
+                }
+                if (nodes[j].id == second_node.id) {
+                    nodes[j].pred.push_back(first_node.id);
+                }
+            }
+        } else {
+            int count_dummy = first_node.layer - (second_node.layer + 1);
+            add_dummy_node(nodes, first_node, second_node, -1);
+            for (int j = 1; j < count_dummy; j++) {
+                add_dummy_node(nodes, nodes.back(), second_node, -1);
+            }
+        }
+    }
+}
+
+void coord_init(std::vector<tree_node> &nodes) {
+    for (int i = 0; i < nodes.size(); i++) {
+        nodes[i].y = (nodes[i].layer + 1) * DISTANCE;
+        nodes[i].x = (nodes[i].number + 1) * DISTANCE;
+    }
+}
+
+void transformation_graph(std::vector<tree_node> &nodes) {
+    std::vector<std::pair<tree_node::node_id, tree_node::node_id>> deleted_edges = {};
+    greedyFAS(nodes, deleted_edges);
+    
+    ASAP(nodes);
+    
+    add_dummy_nodes(nodes, deleted_edges);
+    
+    coord_init(nodes);
+}
+
+const std::vector<tree_node::node_id> Net::get_sources() {
+    std::vector<tree_node::node_id> ids_sources = {};
+    deg_reinit(nodes);
+    for (int i = 0; i < nodes.size(); i++) {
+        if (nodes[i].deg_m == 0) {
+            ids_sources.push_back(nodes[i].id);
+        }
+    }
+    return ids_sources;
+}
+
+const std::vector<tree_node::node_id> Net::get_sinks() {
+    std::vector<tree_node::node_id> ids_sinks = {};
+    deg_reinit(nodes);
+    for (int i = 0; i < nodes.size(); i++) {
+        if (nodes[i].deg_p == 0) {
+            ids_sinks.push_back(nodes[i].id);
+        }
+    }
+    return ids_sinks;
+}
```

Comment:
- Add fields `sources` and `sinks` to the `Net`
- Calculate once
```suggestion
const std::vector<tree_node::node_id> &Net::get_sources() {
    return sources;
}

const std::vector<tree_node::node_id> &Net::get_sinks() {
    return sinks;
}
```